### PR TITLE
Allow override of view definition filenames

### PR DESCRIPTION
**Pull request by bhaibel**

### CODE REVIEW

File: lib/scenic/statements.rb [hunk 1]
```diff
@@ -60,7 +69,7 @@ def create_view(name, version: nil, sql_definition: nil, materialized: false)
     # @example Drop a view, rolling back to version 3 on rollback
     #   drop_view(:users_who_recently_logged_in, revert_to_version: 3)
     #
-    def drop_view(name, revert_to_version: nil, materialized: false)
+    def drop_view(name, revert_to_version: nil, defined_as: nil, materialized: false)
```
Comment (leskat47) [inline lib/scenic/statements.rb]: we're passing in `defined_as`, but not actually using it within the function

**Review by leskat47 — COMMENTED** [review]
My only question is on delete, and I don't see a test for it.

File: lib/scenic/statements.rb (hunk 1 above)
Comment (bethcodes) [inline lib/scenic/statements.rb]: This isn't used in this method?

File: lib/scenic/statements.rb [hunk 2]
```diff
@@ -22,19 +22,28 @@ module Statements
     #     SELECT * FROM users WHERE users.active = 't'
```
Comment (bethcodes) [inline lib/scenic/statements.rb:26]: Update documentation, including the possibility of raising an exception?

File: lib/scenic/statements.rb [hunk 3]
```diff
@@ -88,7 +97,7 @@ def drop_view(name, revert_to_version: nil, materialized: false)
     # @example
     #   update_view :engagement_reports, version: 3, revert_to_version: 2
     #
```
Comment (bethcodes) [inline lib/scenic/statements.rb:107]: ditto on the documentation comment

File: lib/scenic/statements.rb [hunk 4]
```diff
@@ -103,7 +112,16 @@ def update_view(name, version: nil, sql_definition: nil, revert_to_version: nil,
         )
       end
 
-      sql_definition ||= definition(name, version)
+      if defined_as.present? && sql_definition.present?
```
Comment (bethcodes) [inline lib/scenic/statements.rb]: repetitive code: should this be a shared method? or a configuration object? (Not that you need to do that for someone else's code base, but just a thought.)

File: lib/scenic/statements.rb [hunk 5]
```diff
@@ -141,7 +159,9 @@ def replace_view(name, version: nil, revert_to_version: nil, materialized: false
         raise ArgumentError, "Cannot replace materialized views"
       end
 
-      sql_definition = definition(name, version)
+      defined_as ||= name
```
Comment (bethcodes) [inline lib/scenic/statements.rb]: here we don't raise the same error as above: if that is on purpose, perhaps two methods named like `set_definition` and `set_definition!` (is that the Rubyism? I feel like I'm trying to use "extra" correctly at the moment...)

File: lib/scenic/statements.rb (hunk 4 above)
Comment (bethcodes) [reply lib/scenic/statements.rb]: we also use "present" here rather than "blank" above: should these match?

File: lib/scenic/statements.rb [hunk 6]
```diff
@@ -88,7 +97,7 @@ def drop_view(name, revert_to_version: nil, materialized: false)
     # @example
     #   update_view :engagement_reports, version: 3, revert_to_version: 2
     #
-    def update_view(name, version: nil, sql_definition: nil, revert_to_version: nil, materialized: false)
+    def update_view(name, version: nil, sql_definition: nil, defined_as: nil, revert_to_version: nil, materialized: false)
       if version.blank? && sql_definition.blank?
```
Comment (bethcodes) [inline lib/scenic/statements.rb:109]: how does the new case slip through this guard clause?

**Comment by bethcodes** [comment]
I don't find `defined_as` the most explanatory name in the world, but if that logic about `version|sql_definition|defined_as` was centralized in one place the explanation might come from that method name.

File: lib/scenic/statements.rb (hunk 1 above)
Comment (bhaibel) [reply lib/scenic/statements.rb]: They handle migration reversals elsewhere and mirror arguments between the migration and rollback code. (That's also where the `revert_to_version` business comes from.) Didn't want to disrupt the standing pattern.

**Comment by bhaibel** [comment]
@bethcodes I changed `defined_as` to `defined_in` in the hope of greater clarity. I don't want to do too much reorganization of the class before upstream maintainers have a chance to look at the PR -- I'm hoping that being "unobtrusive" will increase my chances of getting the PR into upstream in a timely fashion and reduce the time that we need to rely on an internal fork.

**Review by bethcodes — APPROVED** [review]

**Pull request merged**